File: SiriusJavaClient-Win32/src/main/java/org/sirius/client/win32/classes/DialogBox.java
/**
 * .
 */
package org.sirius.client.win32.classes;

import java.rmi.RemoteException;

import org.sirius.client.win32.Win32Client;
import org.sirius.client.win32.types.Win32Locator;

/**
 * @author Myk Kolisnyk
 * .
 */
public class DialogBox extends TopLevelWindow {

    /**
     * .
     * @param client .
     * @param locator .
     */
    public DialogBox(final Win32Client client, final Win32Locator locator) {
        super(client, locator);
    }

    /**
     * .
     * @param client .
     * @param parent .
     * @param locator .
     */
    public DialogBox(final Win32Client client, final Window parent,
            final Win32Locator locator) {
        super(client, parent, locator);
    }

    @Override
    public final boolean exists() throws RemoteException {
        getLogger().debug("Searching for dialog box: " + getLocator());
        if (getParent() == null) {
            getLogger().debug("Searching for dialog box parent");
            return super.exists();
        } else if (!getParent().exists()) {
            getLogger().debug("Parent window doesn't exist. Quiting...");
            return false;
        } else {
            getLogger().debug("Searching for dialog box in the same thread");
            long hwnd = getClient().utils().searchSameThreadWindow(
                    getParent()
                    .getHwnd(),
                    getLocator());
            if (hwnd == 0) {
                getLogger().debug("No dialog found. Returning false");
                return false;
            } else {
                getLocator().setHwnd(hwnd);
            }
        }
        getLogger().debug("Dialog was found: " + getLocator());
        return true;
    }

    @Override
    public final boolean exists(final long timeout) throws Exception {
        return waitFor(timeout, "exists", true);
    }
}

File: sirius-ruby-client-win32/lib/sirius/win32/classes/DialogBox.rb
module Sirius
  module Client
    module Win32
      module Classes
        class DialogBox < TopLevelWindow
          def initialize(locator,client=nil,parent=nil)
            super(client, parent, locator);
          end

          def exists?()
            if (@parent == nil)
              return super.exists?
            end

            return false unless @parent.exists?

            hwnd = @client.utils.searchSameThreadWindow(
            @parent.hwnd,
            @locator)

            if (hwnd == 0)
              return false
            else
              @locator.hwnd = hwnd
            end
            return true
          end

          #          
          def exists?(timeout)
            return waitFor(timeout, "exists", true)
          end
        end
      end
    end
  end
end